pkgs/build-support/substitute/substitute.nix

```diff
@@ -1,14 +1,54 @@
-{ stdenvNoCC }:
+{ lib, stdenvNoCC }:
+/*
+This is a wrapper around `substitute` in the stdenv.
 
+Attribute arguments:
+- `name` (optional): The name of the resulting derivation
+- `src`: The path to the file to substitute
+- `substitutions`: The list of substitution arguments to pass
+  See https://nixos.org/manual/nixpkgs/stable/#fun-substitute
+- `replacements`: Deprecated version of `substitutions`
+  that doesn't support spaces in arguments.
+
+Example:
+
+```nix
+{ substitute }:
+substitute {
+  src = ./greeting.txt;
+  substitutions = [
+    "--replace"
+    "world"
+    "paul"
+  ];
+}
+```
+
+See ../../test/substitute for tests
+*/
 args:
 
-# This is a wrapper around `substitute` in the stdenv.
-# The `replacements` attribute should be a list of list of arguments
-# to `substitute`, such as `[ "--replace" "sourcetext" "replacementtext" ]`
-stdenvNoCC.mkDerivation ({
+let
   name = if args ? name then args.name else baseNameOf (toString args.src);
+  deprecationReplacement = lib.pipe args.replacements [
+    lib.toList
+    (map (lib.splitString " "))
+    lib.concatLists
+    (lib.concatMapStringsSep " " lib.strings.escapeNixString)
+  ];
+  optionalDeprecationWarning =
+    # substitutions is only available starting 24.05.
+    lib.warnIf (args ? replacements && lib.isInOldestRelease 2405) ''
+      pkgs.substitute: For "${name}", `replacements` is used, which is deprecated since it doesn't support arguments with spaces. Use `substitutions` instead:
+        substitutions = [ ${deprecationReplacement} ];'';
```

Comment:
Remind me: does this mean it will go off for NixOS unstable users who are _already_ in 24.05, or it will go off once 24.11 is cut on (say) May 31st 2024?

Reply:
This `isInOldestRelease` function is super confusing, but I'm pretty sure it's meant to be used when you introduce a new feature (`substitutions` here) to deprecate a feature (`replacements` here).

So the "is in oldest release" terminology refers to the new feature being in the oldest supported release.

If the warning was unconditional, the problem is that third-party projects that aim to support all stable NixOS releases will struggle during the 1 month period in which there's two stable releases, because:
- The new feature isn't usable because the previous release didn't support it
- The old feature would always give a warning and users would complain about this

So the third-party project could either:
- Ignore the user complaints about the warning until the previous release isn't supported anymore anyways
- Implement some logic to use the new feature conditionally on the new release. But this seems barely worth it for that 1 month period.

By using `isInOldestRelease` here, there's no warning for the old feature during the 1 month period where both releases are stable. Therefore no users will be upset.

Though this does push the problem around a bit, because as soon as the previous release isn't supported anymore, the warning will start appearing. So in order to actually avoid warnings for users, the third-party project has to synchronise releases to NixOS, which doesn't seem ideal.

It might be better to have a version of `isInOldestRelease` that is delayed by one more release, such that third-party projects have an entire release cycle to start using the new feature.

Reply:
For what it's worth, I think it's worth sending a message _now_ to users _in 24.05 pre-release_ that they ought to migrate. I find migration _warnings_ that go off many months later to be really surprising and irritating. That's when I expect the _warning_ to transition to an error.

Reply:
I was missing a crucial part, and indeed I think what you're describing is actually what's happening, check out https://github.com/NixOS/nixpkgs/issues/291939#issuecomment-1970073507